aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+// One benefit of the laziness is that we save work in the cases where a user has multiple views and mutates one of them,
+// but never uses the other views later in the program (in which case we'll never update the alias).
+// It also has downsides though: repeatedly applying mutations to the same view withing syncing will slowly leak memory.
+//
+// Corresponding diagram because a picture is 1000 words:
+//
+// b = a.view(...)
+//
+//        a                                                    b
+//        |                                                    |     If the user asks for b and it’s out of date,
+//       \/                                                    \/    We regenerate b by replaying it’s views from the alias.
+// . - - - - - - - - - - - - - .                    . - - - - - - - - - - - - - .
+// |  FunctionalTensorWrapper  |                    |  FunctionalTensorWrapper  |
+// . - - - - - - - - - - - - - .                    . - - - - - - - - - - - - - .
+// |     value   |   storage   |                    |    storage    |   Value   |
+// . - - - - - - - - - - - - - .                    . - - - - - - - - - - - - - .
+//          |                   \                  /                      |
+//          |                     \              /                        |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       |  FunctionalStorageImpl  |           |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       |         Alias           |           |
+//          |                       . - - - - - - - - - - - - .           |
+//          |                       /     mutations to a or b             |
+//          |                     /       are queued onto Alias           |
+//          |                   /                                         |
+//         \/                 /                                           \/
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |        TensorImpl         |                             |             TensorImpl        |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+// |   value   |   storage     |                             |    storage    |     Value     |
+// . - - - - - - - - - - - - - .                             . - - - - - - - - - - - - - - - .
+//          |                                                             |
+//          |                                                             |
+//          |                                                             |
+//          |   In this picture the two tensor views their own storages,  |
+//          |   have their own storages, but backends like functorch      |
+//         \/   are allowed to re-alias underneath the pass               \/
```

Comment:
The idea is that naively for functorch, the functionalization pass would turn this code:
```
b = a.view(...)
b.add_(1)
print(a)
```
into:
```
b = a.view_copy(...)  # view becomes a view_copy (since backends like xla and vulkan don't know how to do view())

b_tmp = b.add(1)
b.replace_(b_tmp)

a_tmp = a.add(1)  # functionalization pass ensure that the mutation also happens to a
a.replace_(a_tmp)
print(a)
```

But functorch doesn't actually want the view_copy: it only wanted mutation removal (which required alias removal). Functorch is run on eager backends like cpu/cuda that know how to run view ops; a view_copy call is actually bad, since it involves creating a new tensor on cpu memory (whereas copies on a backend like xla are cheap, since it just involves copying IR).

So we can add another pass in the future "underneath" the functionalization pass, that will look for every `{view}_copy` operator call, and redispatch instead to a `{view}` call.

I think part of the reason this is confusing is because this note is kinda meant for the final state of the project. Right now, you'll see that the PR doesn't introduce view_copy calls anywhere. That's because it's fusing everything together for now (to get stuff to work for functorch), mostly because we're out of dispatch keys and can't afford 3 separate keys for the alias removal / mutation removal / view_copy->view passes.